87.md

```diff
@@ -0,0 +1,84 @@
+
+This is a different trade-off from NIP 24's encrypted chat, which is simpler and provides more privacy, but is inflexible and doesn't scale as well.
+
+This NIP relies on the following NIPs:
+
+- [86.md](NIP 86) for key sharing
+- [51.md](NIP 51) for member lists
+- [72.md](NIP 72) for community definition
+
+# Protocol Description
+
+## Community Definition
+
+Following NIP 72, communities are identified by the address of a replaceable `kind 34550` event. This may be wrapped using the shared key to support privately-defined communities.
+
+## Admin Key
+
+The key used to publish the community's `kind 34550` definition event MUST be used for performing privileged actions, such as publishing member lists and key rotations. Non-administrative events MAY be published to the community by the admin.
+
+Because the admin key is used for managing key rotation, it's recommended that a dedicated key be used for community administration instead of the community admin's personal key.
+
+## Key Sharing
+
+Key sharing and rotation is accomplished as described in [85.md](NIP 85), with the addition to the `kind 24` event of an `a` tag pointing to the group's address.
+
+## Member Lists
+
+Admins MAY publish member lists using a kind `30000` event with the community's address as the `d` tag, as described in NIP 51. These events MAY be wrapped using NIP 59 to support private member lists.
+
+When rotating the shared key, the admin SHOULD publish an up-to-date member list. If private, the new list SHOULD be addressed to both the old key and the new one individually so clients can update member lists.
+
+## Access Requests
+
+Anyone may request access to a community using a `kind 25` rumor wrapped with an ephemeral key and addressed to the admin's key. The rumor's `content` MAY include a message, and its `tags` MUST include an `a` tag pointing to the community definition event's address.
+
+On receipt, an admin MAY choose to admit the member either by rotating keys or sharing the most recent shared key.
+
+```json
+{
+  "kind": 25,
+  "content": "Pleeease let me in",
+  "tags": [["a", "34550:<admin pubkey>:<group name>"]]
+}
+```
+
+## Exit Requests
+
+Any member may request to be removed from a community using a `kind 25` rumor wrapped with an ephemeral key and addressed to the admin's key. The rumor's `content` MAY include a message, and its `tags` MUST include an `a` tag pointing to the community definition event's address.
+
+On receipt, an admin MAY choose to remove the member either by rotating keys or simply by publishing a new member list.
+
+```json
+{
+  "kind": 26,
+  "content": "I'm outta here",
+  "tags": [["a", "34550:<admin pubkey>:<group name>"]]
+}
+```
+
+## Messages
+
+Any community member may post an event to the community either publicly following NIP 72, or privately. Private events MUST be sealed by the user's own key, and their wrapper signed by
+the shared key. Anyone with the shared key may decrypt these messages.
```

Comment:
Ok, if I need to download all private posts of community `C` what I have to do is: 
1. Download all `kind:24`s with a `a` tag to `C`.
2. Get all public keys of the private keys shared in the `kind:24`. 
3. Create a filter for all GiftWraps that have been authored by all keys from 2. 
4. Use each key to decrypt the GiftWrap. 
5. Check if the sealed event has an `a` tag to `C`. 
6. Display all of 5s events as `C`. 

Is that right? 

That means that we need a way to discard past keys that might have been compromised and were used so fill the community with spam. 

Or should clients only use the last key for each `a`?

Reply:
That's basically right, except step 1 would be kind 24s addressed to the current user (keys are shared with members individually).

When a key is rotated, it invalidates the previous one with a grace period. So there may be multiple keys valid simultaneously, but only for a short period of time. Clients should track that and stop accepting events signed by revoked keys (timestamp spoofing could be done, but that would bury stuff in the timeline, and clients could mitigate that too if needed).

Also, step 4 doesn't require trying all keys, since you know the key the event was encrypted with.